### PR TITLE
feat(enhancement): Distance dependent hidden objects

### LINKED ISSUE #9600: Allow stellar objects to become invisible if you are too far away from them
### Problem Description

As mentioned in #9598, there's currently no way to create hidden objects within a system because the orbit map and radar display everything that a system contains. Even if we could hide an object from those mini-maps, though, objects would still be incredibly easy to find by just zooming out as far as you can and scanning the system visually. 

### Related Issue Links

#9598

### Desired Solution

Create some method by which stellar objects can be made invisible if you are not close enough to them. I will admit that I'm not sure on how the syntax for this would look, as I want this mechanic to be rather customizable.

A checklist of what I'm looking for in this new system is as follows:
* Objects can be given some value where if the player is beyond that distance from the object, the object is invisible, i.e. its alpha is dropped to 0, similar to cloaking ships.
* As you draw closer to the object, the object becomes more visible before returning to full opacity, instead of the above value being a sharp cut-off. Perhaps there is a min and max value where beyond the max it is fully invisible and below the min it is fully visible, with a linear dropoff in opacity between the two.
* Having certain outfits attributes will increase the distance at which you can observe the object.
  * I'm not 100% on using attributes, because then that ties a single attribute to all possible objects we could be trying to hide, when the actual reason/technology behind why an object is hidden may differ between objects. If it is tied to an attribute, then it shouldn't be a hard coded attribute name. Instead, perhaps an object could define the outfit attribute that causes it to become more visible.
* If an object is fully transparent, it can't be selected or landed on. The HUD for the object also won't appear unless the object is landable.
* Visibility of such an object on the radar should still be controlled entirely by #9598.

### Alternative Approaches

N/A

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by tibetiroka**
**Feature**

This is a re-PR of #9602. Fixes #9600.

## Summary

Adds a new attribute to stellar objects that control their visibility.
```
visibility <maxDistance> [<minDistance>]
```
`<maxDistance>` The distance where the object is completely faded out, and cannot be landed on
`<minDistance>` The distance where the object becomes fully visible (optional, defaults to 0)

If minDistance is less than maxDistance, the object disappears as you get closer.

## Screenshots

See the original PR.

## Usage examples
```
system Sol
	[...]
	object Mars
		sprite planet/mars
		distance 1471.43
		period 687.001
		visibility 1000 500
```

## Testing Done
I tested that objects disappear and show up as expected. I tested that the landing thingy doesn't select invisible planets.

NPCs can still land there and take off from invisible planets, and if selected, the selector markers remain visible after the planet disappears. I think this helps players to keep their sanity, but can be removed if necessary.

## Wiki Update
https://github.com/endless-sky/endless-sky-wiki/pull/73

## Performance Impact
There is more transparency going on, but the impact should be minimal.

### CODE REVIEW

File: source/StellarObject.h [hunk 1]
```diff
@@ -72,10 +74,17 @@ class StellarObject : public Body {
 	// Find out how far this object is from its parent.
 	double Distance() const;
 
+	// Change the distances an object is visible from, based on flagship attributes.
```
Comment (TheGiraffe3) [inline source/StellarObject.h:77]: What are these flagship attributes?

Comment (tibetiroka) [reply source/StellarObject.h:77]: Currently, nothing. It's present for future functionality that doesn't have a well-defined syntax yet, because the exact needs aren't completely clear.

File: source/Body.cpp [hunk 2]
```diff
@@ -297,9 +297,27 @@ void Body::SetSwizzle(int swizzle)
...
+
+
+
+double Body::DistanceAlpha(const Point &drawCenter) const
+{
+	if(!distanceInvisible)
+		return 1.;
+	double distance = (drawCenter - position).Length();
+	double distanceAlpha = clamp<double>((distance - distanceInvisible) / (distanceVisible - distanceInvisible), 0., 1.);
+	return distanceAlpha;
```
Comment (TomGoodIdea) [inline source/Body.cpp:313]:
```suggestion
	return clamp<double>((distance - distanceInvisible) / (distanceVisible - distanceInvisible), 0., 1.);
```

File: source/Body.h [hunk 3]
```diff
@@ -77,7 +77,11 @@ class Body {
 	// Set the color swizzle.
 	void SetSwizzle(int swizzle);
 
-	double Alpha() const;
+	// Functions, determining the current alpha value of the body,
```
Comment (TomGoodIdea) [inline source/Body.h:80]:
```suggestion
	// Functions determining the current alpha value of the body,
```

File: source/Body.h [hunk 4]
```diff
@@ -77,7 +77,11 @@ class Body {
...
+	// dependent on the position of the body relative to the center of the screen.
+	double Alpha(const Point &drawCenter) const;
+	double DistanceAlpha(const Point &drawCenter) const;
+	bool IsVisible(const Point &drawCenter) const;
```
Comment (TomGoodIdea) [inline source/Body.h:85]:
```suggestion


```

File: source/Body.h [hunk 5]
```diff
@@ -106,6 +110,9 @@ class Body {
 	float scale = 1.f;
 
 	double alpha = 1.;
+	// The maximum distance at which the body is visible, and at which it becomes invisible again.
```
Comment (TomGoodIdea) [inline source/Body.h:113]: I guess this is for the second variable? If yes, the first one could also use a comment.

**Review by TomGoodIdea — COMMENTED** [review]
I'm not sure about adding these flagship functions that don't have any purpose in this PR

File: source/Body.h (hunk 5 above)
Comment (tibetiroka) [reply source/Body.h:113]: I think this covers both fields?

**Comment by tibetiroka** [comment]
> I'm not sure about adding these flagship functions that don't have any purpose in this PR

The functionality is planned, it just isn't completely clear what syntax/behaviour it should have. @warp-core thoughts?

**Comment by Zitchas** [comment]
The flagship attributes make a lot of sense to me. The instant we have planets, stations, and other sorts of things that are hidden (some intentionally so), is the instant that it also makes sense to have ways to make them visible. Two primary things come to mind:
- For hidden bases, the faction that controls them probably has either a way to provide coordinates to people, or a way to tell members of its faction where the base is. (this would be nullifying the hidden aspect entirely)
- Alternatively, sensors are a thing. Various sensor enhancements could increase the distance at which the player can locate a hidden object.
- And finally, some things I suspect could remain invisible permanently unless the player has something that will make it visible.

In other words:
- A conditional way to make something completely visible.
- A way to increase the sight range of the player.
- A conditional way to make something completely invisible.
- And of course, having these stack. So for instance perhaps the pirate base is totally invisible at any distance unless one has a cryptographic key, in which case it can be revealed at a certain distance. 
- Or perhaps a place is hidden in such a way that normally it can be picked up at a certain distance, but if the player has a certain attribute, it is always invisible and never revealed.

**Comment by TheGiraffe3** [comment]
Question: if an invisible object is a wormhole, will autopilot route through it, or will that part need to be manual?

**Comment by tibetiroka** [comment]
Currently, I believe they would be treated like any other wormhole. This generally means that it is used for route planning if the player knows where it goes, which requires the player to travel through the wormhole first (which can only be done once it's visible).

**Comment by Zitchas** [comment]
Sounds good. My expectation for worm holes would be that if I have travelled through it, and conditions are such that if I approach it it will become visible for me to land at, then my automatic pathfinding should use it.

However, even if I have travelled through it, if conditions are such that when I approach it I won't be able to see it, then automatic pathfinding should not use it.

Of course, this might ruin some element of discovery. What if the wormhole is no longer accessible but there's no good reason the player should know that it isn't accessible anymore? (This isn't a blocking concern, just a thought, and applies just as much to regular wormholes as special hidden ones)

(I think these are both common sense results, but just spelling it out in case we're not on the same page)

File: source/System.cpp [hunk 6]
```diff
@@ -538,6 +539,15 @@ void System::UpdateSystem(const Set<System> &systems, const set<double> &neighbo
 
 
 
+void System::UpdateObjectVisibilities(const Ship *flagship)
```
Comment (petervdmeer) [inline source/System.cpp:542]: This function is currently unused, and we are still discussing if we should run this only on entering a system, or on every frame (with more conditions than only the flagship).

My suggestion would be to leave out this code for now, the code is basically not used in this PR, and when we add modifiers for visibility, then I really hope that we will have more modifiers than only flagship attributes.

File: source/StellarObject.h (hunk 1 above)
Comment (petervdmeer) [reply source/StellarObject.h:77]: As I suggested in another comment: Maybe better to leave those functions out completely, since they are currently not used.

**Review by petervdmeer — COMMENTED** [review]
Code looks good, but the unused functions that are updating game-data are not really nice. (I prefer not to update gamedata on entering of a system, especially not for unused functions.)

**Comment by petervdmeer** [comment]
> The flagship attributes make a lot of sense to me. The instant we have planets, stations, and other sorts of things that are hidden (some intentionally so), is the instant that it also makes sense to have ways to make them visible.

It also makes a lot of sense to me, but I'm still expecting that we can get to an even more flexible system where we can have visibility based not only on flagship attributes, but on anything that we can set in conditions (and preferably full real-time).

Then we can also have options like suddenly seeing the hidden pirate planet once you have captured a derelict ship in-system.

I'm also a big fan of making 2 steps on this topic; Just add distance visibility / invisibility in this PR and have the conditionals for visibility in a follow-up PR. (Then we already get part of the nice feature now, and it allows us to get the other half of the nice feature as complete as possible.)

**Review by petervdmeer — APPROVED** [review]

**Comment by petervdmeer** [comment]
PR looks good. I would be okay with merging this as-is.

File: source/Body.h (hunk 5 above)
Comment (petervdmeer) [reply source/Body.h:114]: It does. Two separate comments might be nice, but I don't think it's a big blocker.

**Comment by petervdmeer** [comment]
Merging: The original PR (#9602) already had quite a number of approvals, and I don't see any strong objections to merging this in either PR discussion thread.
